### DISCUSSION TITLE
Any recommended ways to update the executor of a pre-defined or imported `Electron` object?

### DISCUSSION THREAD

**Discussion by Andrew-S-Rosen** [AgnostiqHQ/covalent#1646] [Q&A]
Is there a convenience function in Covalent to update the executor of a pre-defined Electron? If there's not, would it be worth it to make one?

I know that, for an electron named `myelectron`, you can do something like `myelectron.electron_object.metadata["executor"] = "local"`, but I wasn't sure if that sets all the necessary variables (and if it's recommended).

The reason I'm asking is that I'm developing a library of pre-defined workflows that users will import into their own codes. I'd like to ensure that if a user imports an `Electron` object from my package, they can set the executor to whatever they like. The alternative is for me to define functions and have the user wrap each function in a `ct.electron()` "manually" so they can set the executor when they define the `Electron`, but that's slightly less preferred for my use case.

This is somewhat related to https://github.com/AgnostiqHQ/covalent/discussions/1586 and https://github.com/AgnostiqHQ/covalent/issues/1587, but I've modified the scope of my question a bit.

**Comment by santoshkumarradha** [comment]
hey @arosen93 , this is a long backlogged item on our plate. Ideally the best UX is to have something like 

### UX1
```python
from my_package import energy_electron # imports a direct electron object

energy_electron.executor=MyExecutor()
```

Could @kessler-frost confirm if using `energy_electron.metadata["executor"]=MyExecutor()` would be a valid solution? I'm uncertain if it would properly set everything needed. I recall encountering some issues with UX1 in the past, but perhaps they have since been resolved and we can revisit the matter. 

Meanwhile I am  just curious if using a UX2 would be a good fit for creating custom electron packages.

###  UX2
```python
from my_package import energy_electron
small_cell=energy_electron(**parameters,executor=MyExecutor(num_cpu=2)) # Returns an electron function 
super_cell=energy_electron(**parameters,executor=MyExecutor(num_cpu=45)) # Returns an electron function 

@ct.lattice
def workflow():
    energy_supercell=super_cell("NaCl") #different resource electron
    energy_smallcell=small_cell("NaCl") #different resource electron
```

In certain scenarios, UX2 may be more suitable as it allows for the creation of multiple electron instances with unique parameters through a factory pattern. This is not achievable in UX1, which is designed to return a singleton electron instance with a specific tied instance. Infact our initial interest in UX1 was coming in from a perspective of changing executor inside an interactive notebook after electrons have been defined. 

LMK your thoughts.

**Reply by Andrew-S-Rosen** [reply]
These are both great ideas I think and would address the problem! UX2 is nice in that it allows the user to import the "same" `Electron` multiple times but with multiple different options like you said. That could be very beneficial (your example is exactly what I had in mind). I imagine UX1 you could also do that but you'd have to reset the `.executor` definition between every use where you want to swap it?

Either way, would definitely be very interested to see something like this, as it would allow for much more flexible packages to be built around Covalent!

**Reply by kessler-frost** [reply]
I think we can make both of the UX's work with only minor changes.

**Reply by Andrew-S-Rosen** [reply]
You're really making my day now with all this uplifting news, @kessler-frost!

Issue opened here: https://github.com/AgnostiqHQ/covalent/issues/1647

**Reply by kessler-frost** [reply]
How about a UX2.5 instead of UX2:

```python
from my_package import energy_electron
small_cell=energy_electron(executor=MyExecutor(num_cpu=2))(**parameters) # Returns an electron function 
super_cell=energy_electron(executor=MyExecutor(num_cpu=45))(**parameters) # Returns an electron function 

@ct.lattice
def workflow():
    energy_supercell=super_cell("NaCl") #different resource electron
    energy_smallcell=small_cell("NaCl") #different resource electron
```

This way,
- It will not remove "executor" (or any other field if we decide to add more) from a possible set of argument names
- Will probably be easier to implement

**Reply by kessler-frost** [reply]
Or maybe even swap the positions:

```python
small_cell=energy_electron(**parameters)(executor=MyExecutor(num_cpu=2)) # Returns an electron function 
super_cell=energy_electron(**parameters)(executor=MyExecutor(num_cpu=45)) # Returns an electron function 
```

**Reply by Andrew-S-Rosen** [reply]
I think pretty much any compact solution is probably a good one, so long as it's documented! :D 2.5 (or 2.6 lol) both look good to me.
